hypothesis-python/src/hypothesis/core.py

```diff
@@ -807,7 +807,15 @@ def run_test_with_generator(test):
         if hasattr(test, "_hypothesis_internal_test_function_without_warning"):
             # Pull out the original test function to avoid the warning we
             # stuck in about using @settings without @given.
-            test = test._hypothesis_internal_test_function_without_warning
+            shim = test
+            test = shim._hypothesis_internal_test_function_without_warning
+            # if we set any attributes on the shim test
+            # copy them onto the real test, otherwise we lose them
+            for shim_attribute_name in dir(shim):
+                if "hypothesis" in shim_attribute_name:
+                    setattr(
+                        test, shim_attribute_name, getattr(shim, shim_attribute_name)
+                    )
```

Comment:
I think there's a serious risk of this going wrong as other code changes over time, since it relies on a naming convention that is not explicitly tested - it's very easy to catch too little, or too much.  For example this will copy over the `_hypothesis_internal_test_function_without_warning` attribute too, and so *later* additions will be silently overridden!

With the benefit of hindsight, I think we should rip out the whole 'settings-first adds a shim' thing - while it's helpful in diagnosing misuse, risking bugs like this is much worse.

To compensate, I think we can check for use of settings without given in [our pytest plugin's `runtest_call` hook](https://github.com/HypothesisWorks/hypothesis/blob/master/hypothesis-python/src/hypothesis/extra/pytestplugin.py) (just before the `yield` in the non-hypothesis-test case).  This covers a majority of users much more safely, and is easily extensible to deprecate use of other decorators (such as `@example`, `@seed`, `@reproduce_failure`, etc) without `@given`.

Reply:
I like this better, I've force pushed an update as it's completely different to the original work.

The one downside of this approach is it happens at the `pytest` layer, so the tests for it will be of the form `testdir.runpytest` in a subprocess. I can think of a way round this but not without refactoring the whole codebase.